### DISCUSSION TITLE
SpecDetails in a documentation plugin

### DISCUSSION THREAD

**Discussion by haroon-sheikh** [getgauge/gauge#1836] [General]
Hi All,

I'm currently writing a documentation type plugin similar to spectacle.

Can you please help me understand how `gauge_messages.SpecDetails` are populated when implementing `func (h *handler) GenerateDocs(c context.Context, m *gauge_messages.SpecDetails) (*gauge_messages.Empty, error)`

It's populated when invoking the plugin with the spec path, i.e. `gauge docs 'plugin_name' 'specs_dir'` but I'm trying to figure out how I can get this populated without the specs_dir argument when the plugin is run from the projectroot... `os.Getenv("GAUGE_SPEC_DIRS")` gives the correct project specs path but `m.GetDetails()` comes out as empty.

Any pointers on what I'm missing here?

```go
func (h *handler) GenerateDocs(c context.Context, m *gauge_messages.SpecDetails) (*gauge_messages.Empty, error) {
	allSpecs := GetProtoSpecs(m)
	fmt.Println(len(allSpecs))
	return &gauge_messages.Empty{}, nil
}

func GetProtoSpecs(m *gauge_messages.SpecDetails) []*gauge_messages.ProtoSpec {
	specs := make([]*gauge_messages.ProtoSpec, 0)
	for _, detail := range m.GetDetails() {
		if detail.GetSpec() != nil {
			specs = append(specs, detail.GetSpec())
		}
	}
	return specs
}
```

`GetProtoSpecs(m)` is returning an empty list when spec_dir argument not given on run.

I think the problem is `gauge_specs_dir` default which is being set to relative path `specs`. And running the plugin with relative path argument also returns empty SpecDetails ... e.g. `gauge docs newplugin specs/` fails but `gauge docs newplugin /abs/path/to/specs/` works as expected.

**Comment by haroon-sheikh** [comment]
@BugDiver can you give me hand with this please?

**Reply by haroon-sheikh** [reply]
@sriv Would you have any advise please?

**Reply by sriv** [reply]
Sure @haroon-sheikh - let me dig a little deeper. Looking into this now

**Reply by sriv** [reply]
Some comments:

> I'm trying to figure out how I can get this populated without the specs_dir argument when the plugin is run from the projectroot

`gauge docs` command [accounts for](https://github.com/getgauge/gauge/blob/master/cmd/docs.go#L31) default `specs_dir`. So if that argument is not passed, it should default to `specs`. (I am able to run `gauge docs spectacle` without the specs dir args)

>  please help me understand how `gauge_messages.SpecDetails` are populated

Gauge [populates this](https://github.com/getgauge/gauge/blob/master/plugin/plugin.go#L554-L571) and passes this to the plugin. `SpecDetail` is a struct that contains the information about the spec file, and it's parse errors.

> I think the problem is `gauge_specs_dir` default which is being set to relative path specs. 

`GAUGE_SPECS_DIR` env is [set by gauge to the sources passed in](https://github.com/getgauge/gauge/blob/master/plugin/plugin.go#L392-L397). 

Is you plugin public? If you have your plugin in an installable form, happy to take it for a spin and post back.

**Reply by haroon-sheikh** [reply]
Hi Sriv, thanks for looking into this. You can install the plugin: `gauge install inprogress` https://github.com/sitture/gauge-inprogress

Basically, the problem is when running gauge docs inprogress, the number of specs differ when spec_dir is not passed. 

E.g. 

`gauge docs inprogress` - returns different results to `gauge docs inprogress /abs/path/to/specs`. This is more the case when there's a bigger set of specs.

**Reply by haroon-sheikh** [reply]
Any luck with this @sriv? if you try this in a project with plenty of specs.. you should easily see the differences.

**Reply by sriv** [reply]
Sorry @haroon-sheikh - got held up. I tried this today and I am seeing this error:

```
04-02-2021 17:17:23.756 [inprogress] [CRITICAL] Unable to create data.js file
```

when I run the `inprogress` plugin on https://github.com/getgauge/gauge-tests
